segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/exts3/ExtS3Storage.java

```diff
@@ -0,0 +1,420 @@
+        GetObjectResult<InputStream> result = null;
+        try {
+            result = client.getObject(config.getExts3Bucket(), config.getExts3Root() + streamSegmentName);
+        } catch (S3Exception e) {
+            if ( e.getErrorCode().equals("NoSuchKey")) {
+                result = null;
+            } else {
+                throw e;
+            }
+        }
+        return result != null;
+    }
+
+    private SegmentProperties syncCreate(String streamSegmentName) throws StreamSegmentExistsException {
+        long traceId = LoggerHelpers.traceEnter(log, "create", streamSegmentName);
+
+        if (client.listObjects(config.getExts3Bucket(), config.getExts3Root() + streamSegmentName)
+                  .getObjects().size() != 0) {
+            throw new StreamSegmentExistsException(streamSegmentName);
+        }
+
+        S3ObjectMetadata metadata = new S3ObjectMetadata();
+        metadata.setContentLength((long) 0);
+
+        PutObjectRequest request = new PutObjectRequest(config.getExts3Bucket(),
+                config.getExts3Root() + streamSegmentName,
+                (Object) null);
+
+        AccessControlList acl = new AccessControlList();
+        acl.addGrants(new Grant[]{
+                new Grant(new CanonicalUser(config.getExts3AccessKey(), config.getExts3AccessKey()),
+                        Permission.FULL_CONTROL)
+        });
+
+        request.setAcl(acl);
+
+        client.putObject(request);
+
+        LoggerHelpers.traceLeave(log, "create", traceId);
+        return syncGetStreamSegmentInfo(streamSegmentName);
+    }
+
+    private Void syncWrite(SegmentHandle handle, long offset, InputStream data, int length) throws StreamSegmentSealedException {
+        long traceId = LoggerHelpers.traceEnter(log, "write", handle.getSegmentName(), offset, length);
+
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be read-only.");
+
+        SegmentProperties si = syncGetStreamSegmentInfo(handle.getSegmentName());
+
+        if (si.isSealed()) {
+            throw new StreamSegmentSealedException(handle.getSegmentName());
+        }
+
+        client.putObject(this.config.getExts3Bucket(), this.config.getExts3Root() + handle.getSegmentName(),
+                Range.fromOffsetLength(offset, length), data);
+        LoggerHelpers.traceLeave(log, "write", traceId);
+        return null;
+    }
+
+    private Void syncSeal(SegmentHandle handle) {
+        long traceId = LoggerHelpers.traceEnter(log, "seal", handle.getSegmentName());
+
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be read-only.");
+
+        AccessControlList acl = client.getObjectAcl(config.getExts3Bucket(),
+                config.getExts3Root() + handle.getSegmentName());
+        acl.getGrants().clear();
+        acl.addGrants(new Grant[]{new Grant(new CanonicalUser(config.getExts3AccessKey(), config.getExts3AccessKey()),
+                Permission.READ)});
+
+        client.setObjectAcl(
```

Comment:
In this scenario there are only two possible values. Either `Permission.FULL` or `Permission.READ`.